Add interoperable module guide (#1903)
LiskHQ/lisk-docs: docs/modules/ROOT/pages/build-blockchain/interoperable-module/execute-ccm.adoc

.hello/schemas.ts
[source,typescript]
----
/**
 * Parameters of the reactCrossChain CCM
 */
export interface CCReactMessageParams {
    /**
     * A number indicating the type of the reaction.
     */
    reactionType: number;
    /**
     * ID of the message.
     */
    helloMessageID: string;
    /** Optional field for data / messages. */
    data: string;
}

/**
 * Schema for the parameters of the reactCrossChain CCM
 */
export const CCReactMessageParamsSchema = {
    /** The unique identifier of the schema. */
    $id: '/lisk/hello/ccReactParams',
    type: 'object',
    /** The required parameters for the command. */
    required: ['reactionType', 'helloMessageID', 'data'],
    /** A list describing the available parameters for the command. */
    properties: {
        reactionType: {
            dataType: 'uint32',
            fieldNumber: 1,
        },
        helloMessageID: {
            dataType: 'string',
            fieldNumber: 2,
        },
        data: {
            dataType: 'string',
            fieldNumber: 3,
            minLength: 0,
            maxLength: 64,
        },
    },
};
----

> Create a new file `constants.ts` inside the `hello` module, to store required constants.

Reviewer: ```suggestion
Create a new file `constants.ts` inside the `hello` module, to store the required constants.
```

----
export const CROSS_CHAIN_COMMAND_NAME_REACT = 'reactCrossChain';
export const MAX_RESERVED_ERROR_STATUS = 63;
----

== Create a store for reactions

Create a new store to store the reactions.

.hello/stores/reaction.ts
[source,typescript]
----
import { BaseStore } from 'lisk-sdk';

export interface ReactionStoreData {
    reactions: {
        like: Buffer[];
    };
}

export const reactionStoreSchema = {
    $id: '/hello/reaction',
    type: 'object',
    required: ['reactions'],
    properties: {
        reactions: {
            type: 'object',
            fieldNumber: 1,
            properties: {
                like: {
                    type: 'array',
                    fieldNumber: 1,
                    items: {
                        dataType: 'bytes',
                    },
                },
            },
        },
    },
};

export class ReactionStore extends BaseStore<ReactionStoreData> {
    public schema = reactionStoreSchema;
}
----

== Initialize the cc_command

Initialize a new command `react` with Lisk Commander

[source,bash]
----
lisk generate:command hello react
----

And move it to a new folder `cc_commands` inside the `hello` module.

> Now, open the file `hello/cc_commands/react_command.ts` and import the constants, schemas and types defined above.

Reviewer: ```suggestion
Now, open the file `hello/cc_commands/react_command.ts` and import the constants, schemas, and types defined above.
```

Next, update the following properties of the command:

* Replace `BaseCommand` with `BaseCCCommand`
* Replace `ReactCommand` with `ReactCCCommand`
* Replace `CommandVerifyContext` and `CommandExecuteContext,` with `CrossChainMessageContext`
//TODO: Is it required that the command names for commands creating and executing a specific CCM must match?
* Set the name of the command to `reactCrossChain`.

> * Set the command schema to equal `CCReactMessageParamsSchema`.

Reviewer: ```suggestion
* Set the command schema to match `CCReactMessageParamsSchema`.
```

    const { ccm } = ctx;

    if (ccm.status > MAX_RESERVED_ERROR_STATUS) {
        throw new Error('Invalid CCM status code.');
    }

    const params = codec.decode<CCReactMessageParams>(crossChainReactParamsSchema, ccm.params);
    const messageCreatorAddress = cryptography.address.getAddressFromLisk32Address(
        params.helloMessageID,
    );
    if (!(await this.stores.get(MessageStore).has(ctx, messageCreatorAddress))) {
        throw new Error('Message ID does not exists.');
    }
}
----

Once it is verified that the parameters are valid, we can execute the CCM.

== CCM execution

For this, adjust the `execute()` hook as shown in the snippet below.

The CCM is included in the CCM context `ctx` of the `execute()` hook and can be used to access the CCM parameters.

The <<create-a-store-for-reactions, Reaction Store>> is used to save the reactions for Hello messages.

.hello/cc_commands/react_command.ts
[source,typescript]
----
public async execute(ctx: CrossChainMessageContext): Promise<void> {
		const { ccm, logger } = ctx;
		logger.info('Executing React CCM');
		// const { sendingChainID, status, receivingChainID } = ccm;

> 		// Decode the provided CCM parameters

Author: I added the comments on purpose to provide more context

        const params = codec.decode<CCReactMessageParams>(CCReactMessageParamsSchema, ccm.params);
        logger.info(params, 'parameters');
        // Get helloMessageID and reactionType from the parameters
        const { helloMessageID, reactionType } = params;
        const reactionSubstore = this.stores.get(ReactionStore);
        const messageCreatorAddress = cryptography.address.getAddressFromLisk32Address(helloMessageID);
        let msgReactions: ReactionStoreData;

>         // Get existing reactions for a Hello message, or initialize an empty reaction object, if none exists,yet.

Reviewer: ```suggestion
        // Get existing reactions for a Hello message, or initialize an empty reaction object, if none exists, yet.
```

		try {
			msgReactions = await reactionSubstore.get(ctx, messageCreatorAddress);
		} catch (error) {
			if (!(error instanceof db.NotFoundError)) {
				logger.info({ helloMessageID, crossChainCommand: this.name }, (error as Error).message);
				logger.error({ error }, 'Error when getting the reaction substore');
				throw error;
			}
			logger.info(
				{ helloMessageID, crossChainCommand: this.name },
				`No entry exists for given helloMessageID ${helloMessageID}. Creating a default entry.`,
			);
			msgReactions = { reactions: { like: [] } };

> 		}

Author: This should be updated in the corresponding script first.

		// Check if the reactions is a like

> 		if (reactionType === 0) {

Author: this is something that should be updated in the corresponding script first.

>         // Check if the reactions is a like

Reviewer: ```suggestion
        // Check if the reaction is a like
```

			if (hasLiked > -1) {
				// Remove the sender address from the likes for the message
				msgReactions.reactions.like = msgReactions.reactions.like.splice(hasLiked, 1);
			// If the sender has not liked the message yet
			} else {
				// Add the sender address to the likes of the message
				msgReactions.reactions.like.push(ctx.transaction.senderAddress);
			}
		} else {
			logger.error({ reactionType }, 'invalid reaction type');
		}
		// Update the reaction store with the reactions for the specified Hello message
		await reactionSubstore.set(ctx, messageCreatorAddress, msgReactions);
	}
----

== Final updates of the module and app.ts

Go back to the file `hello/module.ts` and update it as described in the code comments.

.hello/module.ts
[source,typescript]
----
import { BaseInteroperableModule, ModuleMetadata, ModuleInitArgs } from 'lisk-sdk';
import { ReactCrossChainCommand } from './commands/react_command';
import { ReactEndpoint } from './endpoint';
import { ReactMethod } from './method';
import { ReactInteroperableMethod } from './cc_method';
// Import the type for the InteroperabilityMethod
import { InteroperabilityMethod } from './types';

> export class ReactModule extends BaseInteroperableModule {

Author: Replaced all tabs with spaces now, it should have fixed all indentation issues

> public async execute(ctx: CrossChainMessageContext): Promise<void> {
>         const { ccm, logger } = ctx;
>         logger.info('Executing React CCM');
>         // const { sendingChainID, status, receivingChainID } = ccm;
>         // Decode the provided CCM parameters
>         const params = codec.decode<CCReactMessageParams>(CCReactMessageParamsSchema, ccm.params);
>         logger.info(params, 'parameters');
>         // Get helloMessageID and reactionType from the parameters
>         const { helloMessageID, reactionType } = params;
>         const reactionSubstore = this.stores.get(ReactionStore);
>         const messageCreatorAddress = cryptography.address.getAddressFromLisk32Address(helloMessageID);
>         let msgReactions: ReactionStoreData;
> 
>         // Get existing reactions for a Hello message, or initialize an empty reaction object, if none exists,yet.
>         try {
>             msgReactions = await reactionSubstore.get(ctx, messageCreatorAddress);
>         } catch (error) {
>             if (!(error instanceof db.NotFoundError)) {
>                 logger.info({ helloMessageID, crossChainCommand: this.name }, (error as Error).message);
>                 logger.error({ error }, 'Error when getting the reaction substore');
>                 throw error;
>             }
>             logger.info(
>                 { helloMessageID, crossChainCommand: this.name },
>                 `No entry exists for given helloMessageID ${helloMessageID}. Creating a default entry.`,
>             );
>             msgReactions = { reactions: { like: [] } };
>         }
> 
>         // Check if the reactions is a like
>         if (reactionType === 0) {
>             const hasLiked = msgReactions.reactions.like.indexOf(ctx.transaction.senderAddress);
>             // If the sender has already liked the message
>             if (hasLiked > -1) {
>                 // Remove the sender address from the likes for the message
>                 msgReactions.reactions.like = msgReactions.reactions.like.splice(hasLiked, 1);
>             // If the sender has not liked the message yet
>             } else {
>                 // Add the sender address to the likes of the message
>                 msgReactions.reactions.like.push(ctx.transaction.senderAddress);
>             }
>         } else {
>             logger.error({ reactionType }, 'invalid reaction type');
>         }
>         // Update the reaction store with the reactions for the specified Hello message
>         await reactionSubstore.set(ctx, messageCreatorAddress, msgReactions);
>     }

Reviewer: ```suggestion
public async execute(ctx: CrossChainMessageContext): Promise<void> {
    const { ccm, logger } = ctx;
    logger.info('Executing React CCM');
    // const { sendingChainID, status, receivingChainID } = ccm;
    // Decode the provided CCM parameters
    const params = codec.decode<CCReactMessageParams>(CCReactMessageParamsSchema, ccm.params);
    logger.info(params, 'parameters');
    // Get helloMessageID and reactionType from the parameters
    const { helloMessageID, reactionType } = params;
    const reactionSubstore = this.stores.get(ReactionStore);
    const messageCreatorAddress = cryptography.address.getAddressFromLisk32Address(helloMessageID);
    let msgReactions: ReactionStoreData;

    // Get existing reactions for a Hello message, or initialize an empty reaction object, if none exists,yet.
    try {
        msgReactions = await reactionSubstore.get(ctx, messageCreatorAddress);
    } catch (error) {
        if (!(error instanceof db.NotFoundError)) {
            logger.info({ helloMessageID, crossChainCommand: this.name }, (error as Error).message);
            logger.error({ error }, 'Error when getting the reaction substore');
            throw error;
        }
        logger.info(
            { helloMessageID, crossChainCommand: this.name },
            `No entry exists for given helloMessageID ${helloMessageID}. Creating a default entry.`,
        );
        msgReactions = { reactions: { like: [] } };
    }

    // Check if the reactions is a like
    if (reactionType === 0) {
        const hasLiked = msgReactions.reactions.like.indexOf(ctx.transaction.senderAddress);
        // If the sender has already liked the message
        if (hasLiked > -1) {
            // Remove the sender address from the likes for the message
            msgReactions.reactions.like = msgReactions.reactions.like.splice(hasLiked, 1);
        // If the sender has not liked the message yet
        } else {
            // Add the sender address to the likes of the message
            msgReactions.reactions.like.push(ctx.transaction.senderAddress);
        }
    } else {
        logger.error({ reactionType }, 'invalid reaction type');
    }
    // Update the reaction store with the reactions for the specified Hello message
    await reactionSubstore.set(ctx, messageCreatorAddress, msgReactions);
}
```

    public method = new ReactMethod(this.stores, this.events);
    public commands = [new ReactCrossChainCommand(this.stores, this.events)];
    public crossChainMethod = new ReactInteroperableMethod(this.stores, this.events);
    // Create a private member to store the methods of the interoperability module
    private _interoperabilityMethod!: InteroperabilityMethod;

    // ...

    // Assign the methods of the interoperability module to _interoperabilityMethod
    public addDependencies(interoperabilityMethod: InteroperabilityMethod) {
        this._interoperabilityMethod = interoperabilityMethod;
    }

    // Lifecycle hooks
    // eslint-disable-next-line @typescript-eslint/require-await
    public async init(_args: ModuleInitArgs) {
        // Pass the methods of the interoperability module to the reactCrossChain command
        this.commands[0].init({
            interoperabilityMethod: this._interoperabilityMethod,
        });
    }
}
----

Open the `app.ts` file, and register the module to the application.

Because the `ReactModule` is an *interoperable module*, it is required to call `app.registerInteroperableModule()` additionally.

> Last but not least, call the `addDependencies()` method of the `ReactModule` with the methods of the interoperability module as parameter.

Reviewer: ```suggestion
Last but not least, call the `addDependencies()` method of the `ReactModule` with the methods of the interoperability module as a parameter.
```

IMPORTANT: Please remove the redundant registration of the `ReactModule` in the `modules.ts` file. It was added automatically during the command initialization.

.app.ts
[source,typescript]
----
import { Application, PartialApplicationConfig } from 'lisk-sdk';
import { registerModules } from './modules';
import { registerPlugins } from './plugins';
import { ReactModule } from './modules/react/module';

export const getApplication = (config: PartialApplicationConfig): Application => {
    const { app, method } = Application.defaultApplication(config);
    const reactModule = new ReactModule();
    app.registerModule(reactModule);
    app.registerInteroperableModule(reactModule);
    reactModule.addDependencies(method.interoperability);

    registerModules(app);
    registerPlugins(app);

    return app;
};
----

The implementation of a cross-chain command in the Hello module is now completed.

> An example implementation of the interoperable modules React and Hello on two different sidechain is available in the Lisk SDK repository under {url_github_sdk_interop}[examples/interop^]

Reviewer: ```suggestion
An example implementation of the interoperable modules React and Hello on two different sidechains is available in the Lisk SDK repository under {url_github_sdk_interop}[examples/interop^].
Feel free to check it out.
```

.hello/cc_method.ts
[source,typescript]
----
import { BaseCCMethod } from 'lisk-sdk';

export class HelloInteroperableMethod extends BaseCCMethod {}
----

Now, open the `hello/module.ts` file.

* Replace the `BaseModule` class, with `BaseInteroperableModule`.

* Import the `HelloInteroperableMethod` and use it to create a new member `crossChainMethod` inside the module.

In the end, the updated module should look similar to the example below:

.hello/module.ts
[source,typescript]
----
/* eslint-disable class-methods-use-this */
// Replace BaseModule with BaseInteroperableModule
import {
    validator,
    BaseInteroperableModule,
    // ...
} from 'lisk-sdk';
// Import the newly created cc_method
import { HelloInteroperableMethod } from './cc_method';

// ...

export class HelloModule extends BaseInteroperableModule {
       public endpoint = new HelloEndpoint(this.stores, this.offchainStores);
    public method = new HelloMethod(this.stores, this.events);
    public commands = [new CreateHelloCommand(this.stores, this.events)];
       // Assign HelloInteroperableMethod to crossChainMethod

>     public crossChainMethod = new HelloInteroperableMethod(this.stores, this.events);

Author: why 3 dots

    // ...
}
----

Now that the interoperable module is initialized, let's create the cross-chain command which will be used to accept and execute the CCM on the sidechain.

== Add required schemas and constants

Add the schema for the parameters of the React CCM to the `schemas.ts` file inside the `hello` module.

.hello/schemas.ts
[source,typescript]
----

// Schema for the parameters of the crossChainReact CCM

> export const CCReactMessageParamsSchema = {

Author: this is a different module

	// The unique identifier of the schema.
	$id: '/lisk/hello/ccReactMessageParams',
	type: 'object',
	// The required parameters for the CCM.
	required: ['reactionType', 'helloMessageID', 'data'],
	// A list describing the required parameters for the CCM.
	properties: {
		reactionType: {
			dataType: 'uint32',
			fieldNumber: 1,
		},
		helloMessageID: {
			dataType: 'string',
			fieldNumber: 2,
		},
		data: {
			dataType: 'string',
			fieldNumber: 3,
			minLength: 0,
			maxLength: 64,
		},
	},
};
----

Add the corresponding interface to `types.ts`.

.hello/types.ts
[source,typescript]
----
// Parameters of the reactCrossChain CCM

> export interface CCReactMessageParams {

Author: different module

* Set the name of the command to `crossChainReact`.
* Set the command schema to match `CCReactMessageParamsSchema`.

.hello/cc_commands/react_command.ts
[source,typescript]
----
import { BaseCCCommand, CrossChainMessageContext, codec, cryptography, db } from 'lisk-sdk';
import { CCReactMessageParamsSchema, CCReactMessageParams } from '../schemas';
import { MAX_RESERVED_ERROR_STATUS, CROSS_CHAIN_COMMAND_REACT } from '../constants';
import { ReactionStore, ReactionStoreData } from '../stores/reaction';
import { MessageStore } from '../stores/message';

export class ReactCommand extends BaseCCCommand {
    public schema = CCReactMessageParamsSchema;

    public get name(): string {
		return CROSS_CHAIN_COMMAND_REACT;
	}
}
----

== CCM verification

Now, implement the command verification.

To keep the example simple, we only check if the CCM `status` code is valid, and if a Hello message exists for the `helloMessageID` defined in the CCM params.

The CCM to be verified is included in the CCM context `ctx` of the `execute()` hook.

Extend the `verify()` hook to include more checks for the other parameters as well, as desired.

> .hello/cc_commands/react_command.ts

Author: in Hello module the file name is `react_command.ts`

.hello/module.ts
[%collapsible]
====
[source,typescript]
----
import {
	validator,
	BaseInteroperableModule,
	BlockAfterExecuteContext,
	BlockExecuteContext,
	BlockVerifyContext,
	GenesisBlockExecuteContext,
	InsertAssetContext,
	ModuleInitArgs,
	ModuleMetadata,
	TransactionExecuteContext,
	TransactionVerifyContext,
	utils,
	VerificationResult,
} from 'lisk-sdk';
import { CreateHelloCommand } from './commands/create_hello_command';
import { ReactCommand } from './cc_commands/react_command';
import { HelloEndpoint } from './endpoint';
import { NewHelloEvent } from './events/new_hello';
import { HelloMethod } from './method';
import {
	configSchema,
	getHelloCounterResponseSchema,
	getHelloRequestSchema,
	getHelloResponseSchema,
} from './schemas';
import { CounterStore } from './stores/counter';
import { MessageStore } from './stores/message';
import { ReactionStore, reactionStoreSchema } from './stores/reaction';
import { ModuleConfigJSON } from './types';
import { HelloInteroperableMethod } from './cc_method';

export const defaultConfig = {
	maxMessageLength: 256,
	minMessageLength: 3,
	blacklist: ['illegalWord1'],
};

> export class HelloModule extends BaseInteroperableModule {

Author: should be fixed now

	// eslint-disable-next-line @typescript-eslint/require-await
	public async init(args: ModuleInitArgs): Promise<void> {
		// Get the module config defined in the config.json file
		const { moduleConfig } = args;
		// Overwrite the default module config with values from config.json, if set
		const config = utils.objects.mergeDeep({}, defaultConfig, moduleConfig) as ModuleConfigJSON;
		// Validate the provided config with the config schema
		validator.validator.validate<ModuleConfigJSON>(configSchema, config);
		// Call the command init() method with config values as parameters
		this.commands[0].init(config).catch(err => {
			// eslint-disable-next-line no-console
			console.log('Error: ', err);
		});
	}

	public async insertAssets(_context: InsertAssetContext) {
		// initialize block generation, add asset
	}

	public async verifyAssets(_context: BlockVerifyContext): Promise<void> {
		// verify block
	}

	// Lifecycle hooks
	// eslint-disable-next-line @typescript-eslint/require-await
	public async verifyTransaction(_context: TransactionVerifyContext): Promise<VerificationResult> {
		// verify transaction will be called multiple times in the transaction pool
		const result = {

> 			status: 1,

Author: should be updated in script first...

	// eslint-disable-next-line @typescript-eslint/no-empty-function
	public async initGenesisState(_context: GenesisBlockExecuteContext): Promise<void> {}

	// eslint-disable-next-line @typescript-eslint/no-empty-function
	public async finalizeGenesisState(_context: GenesisBlockExecuteContext): Promise<void> {}

	// eslint-disable-next-line @typescript-eslint/no-empty-function
	public async beforeTransactionsExecute(_context: BlockExecuteContext): Promise<void> {}

	// eslint-disable-next-line @typescript-eslint/no-empty-function
	public async afterTransactionsExecute(_context: BlockAfterExecuteContext): Promise<void> {}
}
----
====

Open the `app.ts` file, and register the module to the application.

Because the `HelloModule` is an *interoperable module*, it is required to call `app.registerInteroperableModule()` additionally.

.app.ts
[source,typescript]
----
import { Application, PartialApplicationConfig, NFTModule } from 'lisk-sdk';
import { TestNftModule } from './modules/testNft/module';
import { registerModules } from './modules';
import { registerPlugins } from './plugins';
import { HelloModule } from './modules/hello/module';

> export const getApplication = (config: PartialApplicationConfig): Application => {

Author: should be fixed

            maxLength: 64,
        },
    },
};
----

Add the corresponding interface to `types.ts`.

.hello/types.ts
[source,typescript]
----
// Parameters of the crossChainReact CCM
export interface CCReactMessageParams {
    // A number indicating the type of the reaction.
    reactionType: number;
    // ID of the Hello message being reacted to.
    helloMessageID: string;
    // Optional field for data / messages.
    data: string;
}
----

Create a new file `constants.ts` inside the `hello` module, to store the required constants.

.hello/constants.ts
[source,typescript]
----
export const CROSS_CHAIN_COMMAND_NAME_REACT = 'crossChainReact';
export const MAX_RESERVED_ERROR_STATUS = 63;
----

== Create a store for reactions

Create a new store to store the reactions.

The store will be used in the <<ccm-execution>> to get and set the reactions for a specific Hello message.

In this example, we store the reactions for a Hello message and the sender address of the Hello message, because it is a unique identifier for a Hello message.

.hello/stores/reaction.ts
[source,typescript]
----
import { BaseStore } from 'lisk-sdk';

export interface ReactionStoreData {

>     reactions: {

Author: yeah because we want this example to be extendable with additional reactions types

The store will be used in the <<ccm-execution>> to get and set the reactions for a specific Hello message.

In this example, we store the reactions for a Hello message and the sender address of the Hello message, because it is a unique identifier for a Hello message.

.hello/stores/reaction.ts
[source,typescript]
----
import { BaseStore } from 'lisk-sdk';

export interface ReactionStoreData {
    reactions: {
        like: Buffer[];
    };
}

export const reactionStoreSchema = {
    $id: '/hello/reaction',
    type: 'object',
    required: ['reactions'],
    properties: {
        reactions: {
            type: 'object',
            fieldNumber: 1,
            properties: {
                like: {
                    type: 'array',
                    fieldNumber: 1,
                    items: {
                        dataType: 'bytes',
                    },
                },
            },
        },
    },
};

export class ReactionStore extends BaseStore<ReactionStoreData> {
    public schema = reactionStoreSchema;
}
----

== Initialize the cc_command

Initialize a new command `react` with Lisk Commander

[source,bash]
----
lisk generate:command hello react
----

> To indicate that this command will accept and execute a CCM, move it to a new folder `cc_commands` and rename the file to `react_cc_command.ts`.

Reviewer: ```suggestion
To indicate that this command will accept and execute a CCM, move it to a new folder `cc_commands`, and rename the file to `react_cc_command.ts`.
```

    let { likes } = msgReactions.reactions;
    // Check if the reactions is a like
    if (reactionType === 0) {
        const likedPos = likes.indexOf(senderAddress);
        // If the sender has already liked the message
        if (likedPos > -1) {
            // Remove the sender address from the likes for the message
            likes = likes.splice(likedPos, 1);
            // If the sender has not liked the message yet
        } else {
            // Add the sender address to the likes of the message
            likes.push(senderAddress);
        }
    } else {
        logger.error({ reactionType }, 'invalid reaction type');
    }

    msgReactions.reactions.likes = likes;
    // Update the reaction store with the reactions for the specified Hello message
    await reactionSubstore.set(ctx, msgCreatorAddress, msgReactions);
}
----

== Creating a new endpoint for getting reactions for a Hello message

Last but not least, let's create a new endpoint in the `endpoints.ts` file of the Hello module, to be able to get the reactions for a specific Hello message from the blockchain.

The only required parameter for the request is the sender address of the respective Hello message.

> It is used as unique identifier of a Hello message, to get the corresponding reactions from the <<create-a-store-for-reactions,store>>.

Reviewer: ```suggestion
It is used as a unique identifier of a Hello message, to get the corresponding reactions from the <<create-a-store-for-reactions,store>>.
```
